accepted/2022/ReflectionInvoke.md

```diff
@@ -0,0 +1,354 @@
+Below are some simplfication constraints that may help with any Roslyn implementation around lifetime rules.
+
+**A `TypedReference` can't reference another `TypedReference`**
+```cs
+TypedReference tr = ...
+TypedReference tr2 = __makeref(tr); //  error CS1601: Cannot make reference to variable of type 'TypedReference'
+```
+
+This is similar to trying to make a `TypedReference` to a `Span<int>` above -- both are byref-like types.
+
+However, this limitation for `TypedReference-to-TypedReference` does **not** need to be removed for the goals in this design if it helps with simplifying lifetime rules.
+
+**Just support what is allowed today.** No new capabilities are expected; reflection invokes existing methods which must have already been compiled according to existing rules:
+```cs
+internal class Program
+{
+    static void Main()
+    {
+        // Case 1
+        MyRefStruct rs = default;
+        TypedReference tr = __makeref(rs._span); // Assume we allow instead of CS1601
+        ChangeIt(tr);
+
+        // Case 2
+        Span<int> heap = new int[42];
+        Span<int> stack = stackalloc int[42];
+        CallMe(ref stack, heap); // okay
+        CallMe(ref heap, stack); // CS8350 as expected
+    }
+
+    public static void ChangeIt(TypedReference tr)
+    {
+        // This compiles today:
+        ref Span<int> s = ref __refvalue(tr, Span<int>);
+
+        // And can be assigned to default
+        s = default;
+
+        Span<int> newspan = stackalloc int[42];
+
+        // But this causes CS8352 as expected:
+        s = ref newspan;
+    }
+
+    public static void CallMe(ref Span<int> span1, Span<int> span2) { }
+}
+
+public ref struct MyRefStruct
+{
+    public Span<int> _span;
+}
+```
+
+If allowed today in strongly-typed manner, then reflection should also allow:
+```cs
+internal class Program
+{
+    static void Main()
+    {
+
+        Span<int> span = stackalloc int[42];
+        TypedReference tr = __makeref(span); // Assume we allow instead of CS1601
+
+        // Using a proposed invoke API; calling should be supported passing byvalue
+        MethodInfo mi1 = typeof(Program).GetMethod(nameof(ChangeIt1));
+        mi1.InvokeDirect(target: default, arg1: tr);
+
+        // and supported passing byref
+        MethodInfo mi2 = typeof(Program).GetMethod(nameof(ChangeIt2));
+        mi2.InvokeDirect(target: default, arg1: tr);
+
+        // Just like these methods can be called today:
+        ChangeIt1(span);
+        ChangeIt2(ref span);
+    }
+
+    public static void ChangeIt1(Span<int> span) { }
+    public static void ChangeIt2(ref Span<int> span) { }
+}
+```
+
+FWIW `__arglist` currently compiles with byref-like types:
+```cs
+Span<byte> s1 = stackalloc byte[1];
+Span<byte> s2 = stackalloc byte[11];
+Span<byte> s3 = stackalloc byte[111];
+CallMe(__arglist(s1, s2, s3));
+
+static unsafe void CallMe(__arglist)
+{
+    // However, when enumerating __arglist here, the runtime throws when accessing a byref-like
```

Comment:
Is the runtime tracking lifetime properly in this case?

Reply:
FWIW the compiler does do lifetime tracking here.